modP2pImpl/src/org/aion/p2p/impl1/tasks/TaskSend.java

```diff
@@ -62,89 +66,205 @@ public void run() {
-                                            (ChannelBuffer) attachment,
-                                            this.mgr));
-                        }
-                    }
-                } else {
-                    if (p2pLOG.isDebugEnabled()) {
-                        p2pLOG.debug(
-                                "msg-{} ->{} node-not-exist",
-                                mo.getDest().name(),
-                                mo.getDisplayId());
-                    }
+                    surveyLog.info("TaskSend: wait for message, duration = {} ns.", duration);
+                    wasProcessed = process(mo);
                 }
-                duration = System.nanoTime() - startTime;
-                surveyLog.info("TaskSend: process message, duration = {} ns.", duration);
             } catch (InterruptedException e) {
                 p2pLOG.error("task-send-interrupted", e);
                 return;
             } catch (RejectedExecutionException e) {
                 p2pLOG.warn("task-send-reached thread queue limit", e);
             } catch (Exception e) {
-                if (p2pLOG.isDebugEnabled()) {
-                    p2pLOG.debug("TaskSend exception.", e);
+                p2pLOG.debug("TaskSend exception.", e);
+            }
+        }
+    }
+
+    /**
+     * Returns {@code true} if the message was processed, {@code false} otherwise.
+     *
+     * @return {@code true} if the message was processed, {@code false} otherwise
+     */
+    private boolean process(MsgOut mo) {
+        // shouldn't happen; but just in case
+        if (mo == null) return false;
+
+        // for runtime survey information
+        long startTime, duration;
+
+        startTime = System.nanoTime();
+        // if timeout , throw away this msg.
+        long now = System.currentTimeMillis();
+        if (now - mo.getTimestamp() > P2pConstant.WRITE_MSG_TIMEOUT) {
+            p2pLOG.debug("timeout-msg to-node={} timestamp={}", mo.getDisplayId(), now);
+            duration = System.nanoTime() - startTime;
+            surveyLog.info("TaskSend: timeout, duration = {} ns.", duration);
+            return false;
+        }
+
+        INode node = null;
+        switch (mo.getDest()) {
+            case ACTIVE:
+                node = nodeMgr.getActiveNode(mo.getNodeId());
+                break;
+            case INBOUND:
+                node = nodeMgr.getInboundNode(mo.getNodeId());
+                break;
+            case OUTBOUND:
+                node = nodeMgr.getOutboundNode(mo.getNodeId());
+                break;
+        }
+
+        if (node != null) {
+            // delay sending the request if we have multiple peers and the last sent message was to
+            // the current peer by placing the request in the pendingRequests that will be processed
+            // during the next iteration
+            if (mgr.getActiveNodes().size() > 1 && node.getIdHash() == lastNode.getIdHash()) {
```

Comment:
not a big concern, the size check can just use  " ! isEmpty()"

Reply:
Actually, here the difference is between 1 or larger. If only 1 peer exists, we will not enforce the requirement of not sending another request to the same peer.

Reply:
ah, okay.